ConstraintViolationException while monitoring GTFS-rt feed

**Summary:** 

When the tool is monitoring a GTFS-rt feed, it seems to be failing to insert some warning/error records into the local database.

Here's the stack trace:

~~~
org.hibernate.exception.ConstraintViolationException: could not execute statement
	at org.hibernate.exception.internal.SQLExceptionTypeDelegate.convert(SQLExceptionTypeDelegate.java:59)
	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:42)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:111)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:97)
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:207)
	at org.hibernate.dialect.identity.GetGeneratedKeysDelegate.executeAndExtract(GetGeneratedKeysDelegate.java:57)
	at org.hibernate.id.insert.AbstractReturningDelegate.performInsert(AbstractReturningDelegate.java:42)
	at org.hibernate.persister.entity.AbstractEntityPersister.insert(AbstractEntityPersister.java:2840)
	at org.hibernate.persister.entity.AbstractEntityPersister.insert(AbstractEntityPersister.java:3411)
	at org.hibernate.action.internal.EntityIdentityInsertAction.execute(EntityIdentityInsertAction.java:81)
	at org.hibernate.engine.spi.ActionQueue.execute(ActionQueue.java:619)
	at org.hibernate.engine.spi.ActionQueue.addResolvedEntityInsertAction(ActionQueue.java:273)
	at org.hibernate.engine.spi.ActionQueue.addInsertAction(ActionQueue.java:254)
	at org.hibernate.engine.spi.ActionQueue.addAction(ActionQueue.java:299)
	at org.hibernate.event.internal.AbstractSaveEventListener.addInsertAction(AbstractSaveEventListener.java:318)
	at org.hibernate.event.internal.AbstractSaveEventListener.performSaveOrReplicate(AbstractSaveEventListener.java:275)
	at org.hibernate.event.internal.AbstractSaveEventListener.performSave(AbstractSaveEventListener.java:182)
	at org.hibernate.event.internal.AbstractSaveEventListener.saveWithGeneratedId(AbstractSaveEventListener.java:113)
	at org.hibernate.event.internal.DefaultSaveOrUpdateEventListener.saveWithGeneratedOrRequestedId(DefaultSaveOrUpdateEventListener.java:192)
	at org.hibernate.event.internal.DefaultSaveEventListener.saveWithGeneratedOrRequestedId(DefaultSaveEventListener.java:38)
	at org.hibernate.event.internal.DefaultSaveOrUpdateEventListener.entityIsTransient(DefaultSaveOrUpdateEventListener.java:177)
	at org.hibernate.event.internal.DefaultSaveEventListener.performSaveOrUpdate(DefaultSaveEventListener.java:32)
	at org.hibernate.event.internal.DefaultSaveOrUpdateEventListener.onSaveOrUpdate(DefaultSaveOrUpdateEventListener.java:73)
	at org.hibernate.internal.SessionImpl.fireSave(SessionImpl.java:673)
	at org.hibernate.internal.SessionImpl.save(SessionImpl.java:665)
	at org.hibernate.internal.SessionImpl.save(SessionImpl.java:660)
	at edu.usf.cutr.gtfsrtvalidator.helper.DBHelper.saveError(DBHelper.java:27)
	at edu.usf.cutr.gtfsrtvalidator.background.BackgroundTask.validateEntity(BackgroundTask.java:178)
	at edu.usf.cutr.gtfsrtvalidator.background.BackgroundTask.run(BackgroundTask.java:142)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.sql.SQLIntegrityConstraintViolationException: integrity constraint violation: foreign key no parent; SYS_FK_10137 table: GTFSMESSAGELOG
	at org.hsqldb.jdbc.JDBCUtil.sqlException(Unknown Source)
	at org.hsqldb.jdbc.JDBCUtil.sqlException(Unknown Source)
	at org.hsqldb.jdbc.JDBCPreparedStatement.fetchResult(Unknown Source)
	at org.hsqldb.jdbc.JDBCPreparedStatement.executeUpdate(Unknown Source)
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:204)
	... 31 more
Caused by: org.hsqldb.HsqlException: integrity constraint violation: foreign key no parent; SYS_FK_10137 table: GTFSMESSAGELOG
	at org.hsqldb.error.Error.error(Unknown Source)
	at org.hsqldb.Constraint.getException(Unknown Source)
	at org.hsqldb.Constraint.checkInsert(Unknown Source)
	at org.hsqldb.StatementDML.performIntegrityChecks(Unknown Source)
	at org.hsqldb.StatementDML.insertRowSet(Unknown Source)
	at org.hsqldb.StatementInsert.getResult(Unknown Source)
	at org.hsqldb.StatementDMQL.execute(Unknown Source)
	at org.hsqldb.Session.executeCompiledStatement(Unknown Source)
	at org.hsqldb.Session.execute(Unknown Source)
	... 34 more
~~~

**Steps to reproduce:** 

1. Start the application
1. Enter the following feeds:
    * http://api.tampa.onebusaway.org:8088/trip-updates
    * http://gohart.org/google/google_transit.zip
1. Click Start

**Expected behavior:** 

It should continue to insert error/warnings into the local database without problems

**Observed behavior:** 

It continues to run, but from the stack trace it appears that some records aren't being inserted correctly.

**Platform:** 

Windows 10, jdk1.8.0_112, IntelliJ

--- a/src/main/java/edu/usf/cutr/gtfsrtvalidator/api/model/MessageLogModel.java
+++ b/src/main/java/edu/usf/cutr/gtfsrtvalidator/api/model/MessageLogModel.java
@@ -28,7 +28,7 @@ import javax.xml.bind.annotation.XmlRootElement;
 
 @XmlRootElement
 @Entity
-@Table(name="GtfsMessageLog")
+@Table(name="MessageLog")
 public class MessageLogModel implements Serializable {
 
     public MessageLogModel(){};

--- a/src/main/java/edu/usf/cutr/gtfsrtvalidator/api/model/OccurrenceModel.java
+++ b/src/main/java/edu/usf/cutr/gtfsrtvalidator/api/model/OccurrenceModel.java
@@ -28,7 +28,7 @@ import javax.xml.bind.annotation.XmlRootElement;
 
 @XmlRootElement
 @Entity
-@Table(name="GtfsOccurrence")
+@Table(name="Occurrence")
 public class OccurrenceModel implements Serializable {
     public OccurrenceModel(String elementPath, String elementValue) {
         this.elementPath = elementPath;

--- a/src/main/java/edu/usf/cutr/gtfsrtvalidator/db/GTFSDB.java
+++ b/src/main/java/edu/usf/cutr/gtfsrtvalidator/db/GTFSDB.java
@@ -62,6 +62,8 @@ public class GTFSDB {
         //Use reflection to get the list of rules from the ValidataionRules class
         Field[] fields = ValidationRules.class.getDeclaredFields();
 
+        Session session = InitSessionBeginTrans();
+
         List<ValidationRule> rulesInClass = new ArrayList<>();
         for (Field field : fields) {
             if (Modifier.isStatic(field.getModifiers())) {
@@ -72,7 +74,11 @@ public class GTFSDB {
                         Object value = field.get(rule);
                         rule = (ValidationRule)value;
                         System.out.println(rule.getErrorDescription());
-                        rulesInClass.add(rule);
+                        ValidationRule validationRule = (ValidationRule) session.createQuery("FROM ValidationRule WHERE errorId = "
+                                + "'" + rule.getErrorId() + "'").uniqueResult();
+                        if(validationRule == null) {
+                            rulesInClass.add(rule);
+                        }
                     } catch (IllegalAccessException ex) {
                         ex.printStackTrace();
                     }
@@ -80,8 +86,6 @@ public class GTFSDB {
             }
         }
 
-        Session session = InitSessionBeginTrans();
-        session.createQuery("DELETE FROM ValidationRule").executeUpdate(); //Deleting Error table
         try {
             for (ValidationRule rule : rulesInClass) {
                 session.save(rule);

--- a/src/main/java/edu/usf/cutr/gtfsrtvalidator/helper/DBHelper.java
+++ b/src/main/java/edu/usf/cutr/gtfsrtvalidator/helper/DBHelper.java
@@ -25,7 +25,9 @@ public class DBHelper {
     public static void saveError(ErrorListHelperModel errorListHelperModel) {
         Session session = GTFSDB.InitSessionBeginTrans();
         int messageId = (int) session.save(errorListHelperModel.getErrorMessage());
+        GTFSDB.commitAndCloseSession(session);
 
+        session = GTFSDB.InitSessionBeginTrans();
         for (OccurrenceModel occurrence : errorListHelperModel.getOccurrenceList()) {
             occurrence.setMessageId(messageId);
             session.save(occurrence);

